### ISSUE TITLE
Problem using dask.dataframe.read_sql_table with SQLAlchemy Query object

### ISSUE THREAD

**Issue by aavanian** [dask/dask#2359]
I'm trying to use a SQLAlchemy query statement with the method dask.dataframe and facing the following issue.

Dask seems to be making a limited select from the query passed and that fails.

Sorry, pseudo-code only. It's the same as my case except generic names and I didn't include table definitions. If you can't reproduce easily, I'll complete the code.

```
import dask.dataframe as dd
import pandas as pd
from sqlalchemy.orm.query import Query

q = Query([Table_A.id, Table_A.date, Table_B.name, Table_A.value1, Table_A.value2]).join(Table_B)
uri = 'mysql+pymysql://xxx:xxx@xxx:3306/xxx'

pd.read_sql(q.statement, uri, index_col='id', columns=None)  # OK

dd.read_sql_table(q.statement, uri, index_col='id', npartitions=10, columns=None)  # NOT OK
```

It looks like the query defined in dask/dataframe/io/sql.py line 95:
`q = sql.select(columns).limit(5).select_from(table)`
should be aliasing the subquery.

After a bit of digging, I got it working past that line with:
`q = sql.select(columns).limit(5).replace_selectable(table, table.alias())`
Similar errors happen on further queries in the function (still working on those...) which can be solved with the same modification on each of those queries

I'm no SQLAlchemy expert though so not sure this is the right solution.

Another thing: I noticed that the code line 84-87 of the same file is not behaving as intended (as far as I understand). By then, `columns` is a list of sqlalchemy `Column` objects if `columns` was not passed. From there, the test `if index_col not in columns` will be true if index_col was passed as a string and the index_col may be added again even if it's already there.
It's no problem though as SQLAlchemy somehow ignores the duplicate column (probably because it's not aliased) and write the proper query.

Full stack trace:
```
---------------------------------------------------------------------------
InternalError                             Traceback (most recent call last)
C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\sqlalchemy\engine\base.py in _execute_context(self, dialect, constructor, statement, parameters, *args)
   1181                         parameters,
-> 1182                         context)
   1183         except BaseException as e:

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\sqlalchemy\engine\default.py in do_execute(self, cursor, statement, parameters, context)
    469     def do_execute(self, cursor, statement, parameters, context=None):
--> 470         cursor.execute(statement, parameters)
    471 

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\pymysql\cursors.py in execute(self, query, args)
    165 
--> 166         result = self._query(query)
    167         self._executed = query

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\pymysql\cursors.py in _query(self, q)
    321         self._last_executed = q
--> 322         conn.query(q)
    323         self._do_get_result()

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\pymysql\connections.py in query(self, sql, unbuffered)
    834         self._execute_command(COMMAND.COM_QUERY, sql)
--> 835         self._affected_rows = self._read_query_result(unbuffered=unbuffered)
    836         return self._affected_rows

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\pymysql\connections.py in _read_query_result(self, unbuffered)
   1018             result = MySQLResult(self)
-> 1019             result.read()
   1020         self._result = result

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\pymysql\connections.py in read(self)
   1301         try:
-> 1302             first_packet = self.connection._read_packet()
   1303 

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\pymysql\connections.py in _read_packet(self, packet_type)
    980         packet = packet_type(buff, self.encoding)
--> 981         packet.check_error()
    982         return packet

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\pymysql\connections.py in check_error(self)
    392             if DEBUG: print("errno =", errno)
--> 393             err.raise_mysql_exception(self._data)
    394 

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\pymysql\err.py in raise_mysql_exception(data)
    106     errorclass = error_map.get(errno, InternalError)
--> 107     raise errorclass(errno, errval)

InternalError: (1248, 'Every derived table must have its own alias')

The above exception was the direct cause of the following exception:

InternalError                             Traceback (most recent call last)
<ipython-input-286-c0073156e37e> in <module>()
----> 1 df = dd.read_sql_table(q.statement, uri, index_col='id', npartitions=10, columns=None)

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\dask\dataframe\io\sql.py in read_sql_table(table, uri, index_col, divisions, npartitions, limits, columns, bytes_per_chunk, **kwargs)
     94 
     95     q = sql.select(columns).limit(5).select_from(table)
---> 96     head = pd.read_sql(q, engine, **kwargs)
     97 
     98     if divisions is None:

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\pandas\io\sql.py in read_sql(sql, con, index_col, coerce_float, params, parse_dates, columns, chunksize)
    413             sql, index_col=index_col, params=params,
    414             coerce_float=coerce_float, parse_dates=parse_dates,
--> 415             chunksize=chunksize)
    416 
    417 

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\pandas\io\sql.py in read_query(self, sql, index_col, coerce_float, parse_dates, params, chunksize)
   1082         args = _convert_params(sql, params)
   1083 
-> 1084         result = self.execute(*args)
   1085         columns = result.keys()
   1086 

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\pandas\io\sql.py in execute(self, *args, **kwargs)
    973     def execute(self, *args, **kwargs):
    974         """Simple passthrough to SQLAlchemy connectable"""
--> 975         return self.connectable.execute(*args, **kwargs)
    976 
    977     def read_table(self, table_name, index_col=None, coerce_float=True,

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\sqlalchemy\engine\base.py in execute(self, statement, *multiparams, **params)
   2062 
   2063         connection = self.contextual_connect(close_with_result=True)
-> 2064         return connection.execute(statement, *multiparams, **params)
   2065 
   2066     def scalar(self, statement, *multiparams, **params):

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\sqlalchemy\engine\base.py in execute(self, object, *multiparams, **params)
    943             raise exc.ObjectNotExecutableError(object)
    944         else:
--> 945             return meth(self, multiparams, params)
    946 
    947     def _execute_function(self, func, multiparams, params):

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\sqlalchemy\sql\elements.py in _execute_on_connection(self, connection, multiparams, params)
    261     def _execute_on_connection(self, connection, multiparams, params):
    262         if self.supports_execution:
--> 263             return connection._execute_clauseelement(self, multiparams, params)
    264         else:
    265             raise exc.ObjectNotExecutableError(self)

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\sqlalchemy\engine\base.py in _execute_clauseelement(self, elem, multiparams, params)
   1051             compiled_sql,
   1052             distilled_params,
-> 1053             compiled_sql, distilled_params
   1054         )
   1055         if self._has_events or self.engine._has_events:

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\sqlalchemy\engine\base.py in _execute_context(self, dialect, constructor, statement, parameters, *args)
   1187                 parameters,
   1188                 cursor,
-> 1189                 context)
   1190 
   1191         if self._has_events or self.engine._has_events:

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\sqlalchemy\engine\base.py in _handle_dbapi_exception(self, e, statement, parameters, cursor, context)
   1400                 util.raise_from_cause(
   1401                     sqlalchemy_exception,
-> 1402                     exc_info
   1403                 )
   1404             else:

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\sqlalchemy\util\compat.py in raise_from_cause(exception, exc_info)
    201     exc_type, exc_value, exc_tb = exc_info
    202     cause = exc_value if exc_value is not exception else None
--> 203     reraise(type(exception), exception, tb=exc_tb, cause=cause)
    204 
    205 if py3k:

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\sqlalchemy\util\compat.py in reraise(tp, value, tb, cause)
    184             value.__cause__ = cause
    185         if value.__traceback__ is not tb:
--> 186             raise value.with_traceback(tb)
    187         raise value
    188 

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\sqlalchemy\engine\base.py in _execute_context(self, dialect, constructor, statement, parameters, *args)
   1180                         statement,
   1181                         parameters,
-> 1182                         context)
   1183         except BaseException as e:
   1184             self._handle_dbapi_exception(

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\sqlalchemy\engine\default.py in do_execute(self, cursor, statement, parameters, context)
    468 
    469     def do_execute(self, cursor, statement, parameters, context=None):
--> 470         cursor.execute(statement, parameters)
    471 
    472     def do_execute_no_params(self, cursor, statement, context=None):

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\pymysql\cursors.py in execute(self, query, args)
    164         query = self.mogrify(query, args)
    165 
--> 166         result = self._query(query)
    167         self._executed = query
    168         return result

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\pymysql\cursors.py in _query(self, q)
    320         conn = self._get_db()
    321         self._last_executed = q
--> 322         conn.query(q)
    323         self._do_get_result()
    324         return self.rowcount

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\pymysql\connections.py in query(self, sql, unbuffered)
    833                 sql = sql.encode(self.encoding, 'surrogateescape')
    834         self._execute_command(COMMAND.COM_QUERY, sql)
--> 835         self._affected_rows = self._read_query_result(unbuffered=unbuffered)
    836         return self._affected_rows
    837 

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\pymysql\connections.py in _read_query_result(self, unbuffered)
   1017         else:
   1018             result = MySQLResult(self)
-> 1019             result.read()
   1020         self._result = result
   1021         if result.server_status is not None:

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\pymysql\connections.py in read(self)
   1300     def read(self):
   1301         try:
-> 1302             first_packet = self.connection._read_packet()
   1303 
   1304             if first_packet.is_ok_packet():

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\pymysql\connections.py in _read_packet(self, packet_type)
    979 
    980         packet = packet_type(buff, self.encoding)
--> 981         packet.check_error()
    982         return packet
    983 

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\pymysql\connections.py in check_error(self)
    391             errno = self.read_uint16()
    392             if DEBUG: print("errno =", errno)
--> 393             err.raise_mysql_exception(self._data)
    394 
    395     def dump(self):

C:\Users\alexandre.avanian\AppData\Local\Continuum\Anaconda3\envs\vol_tools3.5\lib\site-packages\pymysql\err.py in raise_mysql_exception(data)
    105         errval = data[3:].decode('utf-8', 'replace')
    106     errorclass = error_map.get(errno, InternalError)
--> 107     raise errorclass(errno, errval)

InternalError: (pymysql.err.InternalError) (1248, 'Every derived table must have its own alias') [SQL: 'SELECT id, date, name, value1, value2 \nFROM (SELECT table_a.id AS id, table_a.date AS date, table_b.name AS name, table_a.value1 AS value1, table_a.value2 AS value2 \nFROM table_a INNER JOIN table_b ON table_b.id = table_a.entity_id) \n LIMIT %(param_1)s'] [parameters: {'param_1': 5}]```

**Comment by martindurant** [comment]
As you can see from the tests, the use of functions and sqlalchemy objects is rudimentary, as I am no expert here either (this is the reason for calling it read_sql_table, because there is a limit to how general it can be). If your change can enable a wider range of expressions without breaking the existing ones, I'd be more than happy to see it! You are welcome to submit a PR.

Having said that, the read_sql code is phrased in terms of the SQL expressions language, not the ORM, does `q.statement` produce an expression or a string?

> Another thing: I noticed that the code line 84-87 of the same file is not behaving as intended (as far as I understand)

I think this is more a lack of documentation: if you are using SQLalchemy objects for the query, then index_col should also be an object, one that's in the query. You are right that strings and objects don't mix here; or did I misunderstand you?

**Comment by aavanian** [comment]
`q.statement` produces a `sqlalchemy.sql.annotation.AnnotatedSelect` which is a `Selectable` so a priori accepted by `pandas.read_sql` as per the doc.
I'll need a bit of time to check this change doesn't break the other ways to pass `table` (from the way the function works, it should be ok, but...)

Understood for `index_col`, that makes sense.

**Comment by aavanian** [comment]
So I wrote two new tests for that case and ... they pass ... because it's dependent on the sql dialect. I rewrote them to test both sqlite and mysql and they fail for mysql on a test db.
With my fix, all tests (the two new and the existing ones) pass on both dialects except `test_with_func` so I will need to further check this one.
You can check [this branch](https://github.com/aavanian/dask/tree/2359-read_sql_table-from-sqla-selectable) for the changes. No PR yet as several choices about test mysql are not straightforward (do you want to? how to configure the db? etc...)

I also had a quick look at how Pandas tests different dialects and let's say it much beyond my knowledge or the time I can commit to this.

**Comment by aavanian** [comment]
I was also continuing my testing and discovered that `sqlalchemy.sql.annotation.AnnotatedSelect` objects are not pickable by cloudpickle (neither by the vanilla pickle).

SQLAlchemy provides `dumps`/`loads` functions in `sqlalchemy.ext.serializer` that enable pickling of objects of this family so I suppose cloudpickle or dask could be adapted to handle this but not sure where/how.

If such expressions can't be used with dask-distributed, it doesn't make much sense to add support for them.
Let me know your thoughts re: path forward on this and I'll see what I can do. In the meantime, I think I'll use delayed to perform the load (there's such an example in the docs).

**Comment by martindurant** [comment]
It certainly is possible with to use sqlalchemy's functions as [custom serialization](http://distributed.readthedocs.io/en/latest/serialization.html) with the distributed scheduler. 

I am assuming this may be an issue with `delayed` too (if your inputs are sqlalchemy select objects). Is this all because of the ORM being tied to a specific database instance? You mentioned SQL dialect-dependent behaviour, which surprised me, but this must be the reason (I expect the rawer sqlalchemy expressions syntax to be more robust).

In short, if you have success building a delayed pipeline, we can work at making it more general in this module. I doubt read_sql_table (nor pandas' version!) gets much use at the moment, and it really was written with the simple case in mind and the functions/expressions capabilities as bonuses. You'll notice that the sql writer counterpart doesn't exist yet, although that should be simpler logic.

**Comment by aavanian** [comment]
I couldn't make `register_serialization` work. Same error as before and the stack trace doesn't show it going through the custom serialization functions I register.

I made it work through `delayed`. No seralization issue there as the `Query` object is not created yet (dask serialize the function that will create it). And as we use then use pandas to read the `Query` object, we don't have further issue (i.e. works out of the box).

sample code (can be used as a recipe):
````
def find_nrecords():
    with session_scope() as session:
        q = session.query(func.count(Table_A.id)).one()
    return q[0]

def load_partition(limits):
    offset, limit = limits
    uri = 'mysql+pymysql://xxx:xxx@xxx:xxx/xxx'
    q = Query(...).join(...).limit(limit).offset(offset)
    df = pd.read_sql(q.statement, uri, index_col='...')
    return df

def load_partitions(npartitions=10):
    nrec = find_nrecords()
    per_part = nrec // npartitions
    per_part += 1 if nrec - npartitions * per_part > 0 else 0  # Add one more to ensure covering the remainder
    args = [(i * per_part, per_part) for i in range(npartitions)]
    delayeds = [delayed(load_partition)(arg) for arg in args]
    return dd.from_delayed(delayeds)
````

The alternative, to make it work from `dask.dataframe.read_sql_table`, two fixes are needed:
* the way the queries are reworked inside that function (work-in-progress, see link above)
* make the custom serialization work (I'll give it another try)

**Comment by mrocklin** [comment]
One might also ask the question of why the SQLAlchemy query isn't
serializable.  Is this true of all sqlalchemy objects?  Perhaps this
contains a reference to a live database connection and shouldn't be
serialized?  If not then perhaps this can be fixed upstream in the
sqlalchemy project itself.

On Sun, May 21, 2017 at 1:49 AM, Alexandre Avanian <notifications@github.com

**Comment by aavanian** [comment]
Here's the link to the [relevant](http://docs.sqlalchemy.org/en/latest/core/serializer.html) part of the docs. It's not related to any actual connection. The project provide a serializer/deserializer for this, it does work on the query object I'm using but I didn't manage to make it work with `register_serialization`: it doesn't seem to be called and the regular pickle/cloudpickle fail on the object. I'll give it another try tomorrow.

**Comment by mrocklin** [comment]
Long term we might ask the sqlalchemy project to implement the
`__getstate__`  `__setstate__` protocols to connect their serialization
functions to normal pickle operation.

On Sun, May 21, 2017 at 9:11 AM, Alexandre Avanian <notifications@github.com

**Comment by aavanian** [comment]
So:
* regarding my fix for sqla selectables which was still failing on `test_with_func`: I have a solution [there](https://github.com/aavanian/dask/commits/2359-read_sql_table-from-sqla-selectable) Let me know if you want a PR for that.
* regarding SQLA serialization of selectable, I found a few discussions and it seems quite unsupported/fragile. In particular: https://bitbucket.org/zzzeek/sqlalchemy/issues/3918/attributeerror-on-self_hash-when
* that being said, I made the registration of the custom serializer work on the initial query but the SQLAlchemy serializer still fails on the updated query that `dask.dataframe.read_sql_table` produces. It's possible the fix mentioned in the sqlalchemy issue linked above works here but I don't have time to test it.

**Comment by martindurant** [comment]
Thanks for looking into this, @aavanian . I'm afraid I don't have much of a solution for you - perhaps constructing your sqlalchemy expressions in delayed functions is indeed the best course for you. It looks unlikely that dask.dataframe.read_sql_table can be forces into accepting every type of input (without serialization issues).

**Comment by aavanian** [comment]
I tend to agree. Even if fixed, it would probably still be quite fragile.
In any case the delayed approach works perfectly (still a small thing to investigate: why persisting the resulting dataframe doesn't return immediately).
